### DISCUSSION TITLE
基于vite的vue-mini打包工具

### DISCUSSION THREAD

**Discussion by FliPPeDround** [vue-mini/vue-mini#105] [Show and tell]
https://github.com/FliPPeDround/vite-plugin-vue-mini
 #93 #65

基于vite的一个打包工具
支持ts及样式预处理器

使用方法
```ts
import { defineConfig } from 'vite'
import Vmini from 'vite-plugin-vue-mini'

export default defineConfig({
  plugins: [
    Vmini(),
  ],
})
```

**Comment by yangmingshan** [comment]
不明觉厉！

**Comment by FliPPeDround** [comment]
只实现了基本的功能，目前还有一些问题，如果你觉得这是一个可行的方向话，我们可以一起做得更好

**Reply by FliPPeDround** [reply]
因为我看作者在其他的回复中，并不认为使用vite/webpack对于vue-mini来说是一个更好的解决方案，所以我对于要不要继续完善下去，有些犹豫
但我认为使用vite的好处有很多
1. 看起来更专业
2. 支持样式预处理器
3. 支持ts
4. 因为plugin的原因，更方便做功能的扩展
5. 未来rolldown落地后会很快
6. DX会更好

**Reply by yangmingshan** [reply]
我十分粗略的看了一下，还是没太理解。vite 在这其中的角色是什么？我看文件也都是你自己扫描的，组件库也是自己拷贝的。vite 干了什么？只负责文件的 transform？

**Reply by FliPPeDround** [reply]
目前来说是这样，因为小程序的路径寻址，是通过json配置的，我没想到太好的办法在vite中实现
但是，如果后续要实现sfc或者一些其他需要编译的能力时，才能体现vite的价值
所以是更方便做功能的扩展

**Reply by yangmingshan** [reply]
其实这就是我一直以来怀疑的地方，vite 核心的功能小程序用不上。

我理解 vite 主要是两部分，dev server + production bundler。dev server 用于响应浏览器的 import 请求，这个小程序自然用不上。production bundler 也用不上，因为小程序不需要打包，至少不需要开发者自己打包，上传小程序时微信开发者工具或者 CLI 会做打包，但这个超出了开发者的控制。

小程序有着自己的一套运行时规则，跟 web 不兼容，连 npm 都是自己搞了一套。小程序不需要 dev server 也不需要 production bundler，它需要的是一对一的源文件 transform + node_modules 到 miniprogram_npm 的搬运 / 映射。小程序需要专门为它量身定制的工具，一个面向小程序的 vite。

我觉得你的项目很棒，并没有任何别的意思。但是从我觉得 vite 在其中存在的必要性很弱，完全可以把 vite 这个壳拿掉，直接把文件输送给 babel postcss 等工具处理，最终效果是一样的，因为 vite 内部也是调用这类工具。这其实就是 create-vue-mini 目前的做法。

其实 create-vue-mini 的远期目标就是发展成一个小程序版的 vite。核心功能会做封装，会留出扩展接口，会设计插件系统等等。

不过我不是 vite 专家，我的理解可能不准确。也许 vite 确实足够灵活能改造成理想的小程序工具，所以我觉得大家的各种探索都是好事。说不定哪天我就被打脸了。

**Reply by FliPPeDround** [reply]
我能理解你的意思，
或许我和一些用户对于`vue-mini`未来的期许是一个需要打包的框架

但不太明白你看起来有些排斥（可能`排斥`不太准确；））vite的原因
正如你所说，样式预处理器自己用postcss也能做，那为什么交给vite处理就是一个不合适的方式
ts文件同理

`原因其实 create-vue-mini 的远期目标就是发展成一个小程序版的 vite。核心功能会做封装，会留出扩展接口，会设计插件系统等等。`对于此，我认为是非常好的计划，但我对于一个全新打包工具的插件接口的设计和是否有开发人员愿意学习一套只在vue-mini里有用的技术持怀疑态度

`也许 vite 确实足够灵活能改造成理想的小程序工具`我认为vite提供的灵活性没有任何问题和障碍

最后感谢能这么晚了，还回答我的疑问
晚安，好梦

**Comment by Airkro** [comment]
之前的交流我就发现，好像双方是在各说各的，但现在觉得有点清楚作者的意思了，让我从几个方面讲一下我个人更喜欢用打包工具（我认为也是为什么Taro/MpVue/MPX/等等都用打包工具）：

**首先，小程序自己 transform 的能力，够用吗？**

1. 小程序自身的编译能力并不完美，只支持非常过时的 ES5 -> ES6，不支持大量ES新语法，也没有好的 Polyfill。
2. 小程序自带的 npm 支持也相当落后，不支持社区的新路径规范，编译能力更差，由于先打包一遍，tree shaking 效果不佳。
3. 小程序自带的编译能力不支持自定义文件打包，各类定制化语法的打包，打包出错也很难排查错误。
4. 小程序开发者工具的编译，和小程序官方提供的 CI 打包工具，打出来的代码是不一致的，这加大了风险。

因此，我并不觉得小程序自带的打包能力“够用”，光是为了解决这些问题就值得用打包工具。

小结：

- 自带的 transform 是不够强的，问题一大堆。
- 原生 npm 编译不好用，如果要用 npm 的包，仅仅“搬运”和“一对一 transform ”是不够的。因此“打包”是需要的。
- 用了打包工具，“编译”能增强，“打包”是额外送的，“搬运”也能胜任。

在满足了 transform 和 npm 支持的基本条件外，打包工具还送全家桶：

- 在已经使用了打包工具的基础上，想要 TypeScript/Sass/Less/Tailwind/UnoCSS 什么的都是直接复用社区即可，不用人操心的。
- 除开打包能力，vite/webpack/rollup 作为扩展功能的基座，它们的生态丰度是无可比拟的，其他的竞争对手都被熬死了。
- 用了打包工具，就意味着打开了和 H5 跨端复用代码的大门，最终能否做到先不说，但至少迈进了第一步，使用了通用的基建。

**Comment by Airkro** [comment]
作者开发 create-vue-mini 是值得肯定的，但正如我想的一样，社区很快就会基于 vite/webpack/rspack 搞出五花八门的编译实现，而且以前也存在，只是没有明星项目而已。

虽然 create-vue-mini 也打算提供各种各样的功能，但恕我直言，不仅要新学一套工具，自定义配置容易不一致，**尤其是想要新功能还要等**。这涉及到信心和耐心的问题。

当然了，作者绝对有权力继续深耕 create-vue-mini 。我只是对“小程序不需要打包”这一言论持反对意见。

**Comment by Airkro** [comment]
但我也有一个问题：

其实现阶段的 [vue-mini](https://github.com/vue-mini/vue-mini) 和 打包工具其实可以完全不相干，用 vue-mini 的项目可以不用任何工具，用了编译打包工具的项目其实也可以不用 vue-mini。

所以打包工具真的需要进 Roadmap ，由作者带头实现吗？社区自己就能做啊。

**Reply by yangmingshan** [reply]
当然可以，十分欢迎👏

**Comment by yangmingshan** [comment]
@FliPPeDround @Airkro 如果你们相信 Vite 是正确的方向，我觉得你们应该坚持做下去，探索你们的解决方案，不必在意我的想法。同人打败官方是很正常的一件事，Vue Mini 也是走在这条道路上。我很乐意帮忙宣传你们的成果。

**Reply by FliPPeDround** [reply]
感谢支持，但我认为现在的vue-mini对于是否使用vite作为打包工具的必要性并不大
正如你所说，vite能做的只是将一些工具的整合，即使不用vite也能实现相同的功能，尽管配置要复杂一些
如果vue-mini暂时的计划是只作为一个微信小程序的扩展库，不打算做一些编译时的功能，提供一些框架能力时，我也认为`build.js`是目前的比较好的解决方案，毕竟对于用户来说不是黑盒
所以我个人认为，如果不能一起推进vue-mini提供编译时功能(比如`sfc`, `defineOptions`, `app.config.js`)，`vite-plugin-vue-mini`这类库能做的事情很少，继续改善的意义不大。仅个人观点，如果有人原因继续完善，个人愿意提供力所能及的帮助

**Comment by FliPPeDround** [comment]
今天看到[weapp-vite](https://github.com/sonofmagic/weapp-tailwindcss/tree/main/packages/weapp-vite)一个基于vite的微信小程序打包工具，或许可以直接配合`vue-mini`使用

**Reply by watsonhaw5566** [reply]
是的，之前也有在关注到，是 weapp-tailwind 的作者搞的。

你的思路很不错。
